### ISSUE TITLE
3p-cookies/step2.html does not have the frame-ancestors HTTP Header set in HTTP response for Admin UI

### ISSUE THREAD

**Issue by raoganeshr** [keycloak/keycloak#10538]
### Describe the bug

Most of the pages in Keycloak Admin UI has the frame-ancestors property set (frame-src 'self'; frame-ancestors 'self'; object-src 'none';). However auth/realms/master/protocol/openid-connect/3p-cookies/step2.html does not have frame-ancestors 'self' it seems. 

Is this as designed or a miss?

### Version

16.1

### Expected behavior

Expect the response header too has frame-ancestors 'self'; set.

### Actual behavior

Response header does not have frame-ancestors value.

### How to Reproduce?

Navigate to auth/realms/master/protocol/openid-connect/3p-cookies/step2.html and observe the response headers in Chrome Developer tools.

**Comment by stianst** [comment]
Could you explain when this is an issue? The steps you have listed in how to reproduce mentions opening this page directly, but you are not supposed to open this page directly.

**Comment by raoganeshr** [comment]
Hi Stian,

The url came up as a potential issue due a Qualys scan of the master realm. The HTTP response for this page does not return a frame-ancestors directive as other keycloak pages do (with default keycloak settings.)

**Comment by opdt** [comment]
@stianst This issue prevents us from using a dedicated admin hostname. The admin console seems to be waiting on the embedded iframe to load _/realms/master/protocol/openid-connect/3p-cookies/step1.html_ and just displays a loading spinner.
This message is logged on the client side: _Timeout when waiting for 3rd party check iframe message._

Since this prevents us to deploy Keycloak in a secure way (dedicated admin host), could you please reconsider the priority of fixing this issue?

**Comment by ahus1** [comment]
@stianst @mposolda - there is now PR #20932 to fix this

**Comment by pedroigor** [comment]
@opdt In your case, the `X-Frame-Options` is kinda causing this behavior, no? By setting `SAMEORIGIN`, you won't be able to frame pages from Keycloak.

**Comment by opdt** [comment]
As i said, our infra re-adds the X-Frame-Options header with value `SAMEORIGIN` for security reasons. Having `frame-ancestors *` explicitly in the CSP makes it work again (we override the CSP via our own proxy inside Kubernetes right now, so this is already verified).

**Comment by tsaarni** [comment]
I cannot really speak on behalf of the original submitter, but as far as I understood, the issue is that security scanners report the pages as potentially vulnerable:

@raoganeshr wrote:
> The url came up as a potential issue due a Qualys scan of the master realm. The HTTP response for this page does not return a frame-ancestors directive as other keycloak pages do (with default keycloak settings.)

This may be false positive from the scanner. Regardless, it would be nice if even false positives can be removed, if possible. 

From that perspective, would `frame-ancestors *` help with the original report or is it fixing another issue of Keycloak being flagged for potential security issue?

**Comment by pedroigor** [comment]
Thanks. If we can remove the false positive better, I agree.

For that particular page, we are supposed to allow any ancestor. This is explicitly set in the code. Not sure if the intent was to rely on browser defaults that at the time the code was written were allowing any ancestor if the header wasn't set. I don't know.

But for this particular case, `self` won't work as it will allow only the server to embed the page. I can't think of any reason for not setting `frame-ancestors *` if that is how we are supposed to indicate to browsers that those pages can be embedded by other origins.

**Comment by raoganeshr** [comment]
I agree. if we need to allow everyone to load it, explicitly setting frame-ancestors to * will denote its intended instead of not setting it which implies that it could be a miss.

**Comment by LakshmiPhani7680** [comment]
Hi @tsaarni  or @pedroigor  or @mposolda  or @stianst  or @ahus1, where is this step1.html or step2.html, how can i add the X-Frame-Options header to this url https://mykeycloak/auth/realms/master/protocol/openid-connect/3p-cookies/step1.html? Kindly respond ASAP as a vulnerability got reported so called Clickjacking-Frameable.
Thank you

**Comment by jonkoops** [comment]
Closing this issue in favor of #29782
